internal/broker/broker.go

```diff
@@ -396,11 +402,49 @@ func (b *Broker) IsAuthenticated(sessionID, authenticationData string) (string,
 	return access, data, nil
 }
 
-func (b *Broker) handleIsAuthenticated(ctx context.Context, session *sessionInfo, authData map[string]string) (access, data string) {
+type isAuthenticatedResult struct {
+	access   string          `json:"-"`
+	UserInfo json.RawMessage `json:"userinfo,omitempty"`
+	Message  string          `json:"message,omitempty"`
+}
+
+func (iar *isAuthenticatedResult) encode() (string, error) {
+	if iar.UserInfo == nil && iar.Message == "" {
+		return "", nil
+	}
+	b, err := json.Marshal(iar)
+	if err != nil {
+		return "", err
+	}
+	if testing.Testing() {
+		// Indent in testing, for improve readability of golden files
+		var indented bytes.Buffer
+		err = json.Indent(&indented, b, "", "  ")
+		if err != nil {
+			return "", err
+		}
+		b = indented.Bytes()
+	}
+	return string(b), err
+}
+
+func (iar isAuthenticatedResult) result() (string, string) {
+	ret, err := iar.encode()
+	if err != nil {
+		slog.Debug(fmt.Sprintf("Error when encoding %#v: %v", iar, err))
+		return iar.access, ""
+	}
+	return iar.access, ret
+}
```

Comment:
This wrapping is not necessary. Since the `access` and `data` only matter when going out of `IsAuthenticated`, I would replace the places where we need to unwrap this to use `isAuthenticatedResult` instead and then handle this logic at the end of the `IsAuthenticated` function (especially since, if we fail to encode the data, we need to return either `Retry` or `Denied`).

Reply:
It was  not to repeat the same in 3 places, but ok...